### PR TITLE
OFI: Simplify random STX allocator

### PR DESCRIPTION
**Pull request by jdinan**
Signed-off-by: James Dinan <james.dinan@intel.com>

### CODE REVIEW

**Review by davidozog — CHANGES_REQUESTED** [review]
Yeah, I like it and it's much simpler.

When I check the debug output this appears to do a round-robin allocation of STXs.  I think it's because each new allocation does a linear search from the unused pool, which sequentially picks the next STX.  This may have come up before, but I can't imagine that's the behavior we want out of the random allocator...

**Comment by jdinan** [comment]
If I coded it right, it should choose at random from the shared STX pool, but sequentially from the free STX pool.  Choosing randomly from the free STX pool doesn't seem like it will make a difference and will introduce a significant amount of complexity.

**Comment by davidozog** [comment]
What I'm seeing looks like there is only 1 STX in the shared pool that is ever below the threshold.  That one gets filled up before selecting an unused one.  Could that be a small mistake in the logic somewhere?

**Comment by davidozog** [comment]
Maybe the output will clarify what I mean.  I might be making a mistake too:

```
OMP_NUM_THREADS=64 SHMEM_OFI_STX_THRESHOLD=8 SHMEM_OFI_STX_DISABLE_PRIVATE=1 SHMEM_OFI_STX_ALLOCATOR=random SHMEM_OFI_STX_MAX=8 SHMEM_DEBUG=1 ./shmem_ctx 
                                                                                                                                                                                     
travis@a5e628e89608:~/SOS-jim_rand/test/unit$  OMP_NUM_THREADS=64 SHMEM_OFI_STX_THRESHOLD=8 SHMEM_OFI_STX_DISABLE_PRIVATE=1 SHMEM_OFI_STX_ALLOCATOR=random SHMEM_OFI_STX_MAX=8 SHMEM_DEBUG=1 ./shmem_ctx 

(grep STX)
                                                                                                                                                                                                     
[0000]        STX[8] = [ 1S 0S 0S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 2S 0S 0S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 3S 0S 0S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 4S 0S 0S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 5S 0S 0S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 6S 0S 0S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 7S 0S 0S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 8S 0S 0S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 9S 0S 0S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 9S 1S 0S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 9S 2S 0S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 9S 3S 0S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 9S 4S 0S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 9S 5S 0S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 9S 6S 0S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 9S 7S 0S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 9S 8S 0S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 9S 9S 0S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 9S 9S 1S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 9S 9S 2S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 9S 9S 3S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 9S 9S 4S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 9S 9S 5S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 9S 9S 6S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 9S 9S 7S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 9S 9S 8S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 9S 9S 9S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 9S 9S 9S 1S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 9S 9S 9S 2S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 9S 9S 9S 3S 0S 0S 0S 0S ]
...
```

**Comment by davidozog** [comment]
Sorry for the super long debug posts... 

The original code has somewhat similar behavior at first, but (kinda) does what I'd expect on STX's 2 and 3 (0-indexed).  But it still doesn't look quite right to me...

```
[0000]        STX[8] = [ 1S 0S 0S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 2S 0S 0S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 3S 0S 0S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 4S 0S 0S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 5S 0S 0S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 6S 0S 0S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 7S 0S 0S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 8S 0S 0S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 9S 0S 0S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 10S 0S 0S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 10S 1S 0S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 10S 2S 0S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 10S 3S 0S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 10S 4S 0S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 10S 5S 0S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 10S 6S 0S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 10S 7S 0S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 10S 8S 0S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 10S 9S 0S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 10S 9S 1S 0S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 10S 9S 1S 1S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 10S 9S 1S 2S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 10S 9S 1S 3S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 10S 9S 1S 4S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 10S 9S 1S 5S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 10S 9S 1S 6S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 10S 9S 1S 7S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 10S 9S 1S 8S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 10S 9S 1S 9S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 10S 9S 2S 9S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 10S 9S 3S 9S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 10S 9S 4S 9S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 10S 9S 5S 9S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 10S 9S 6S 9S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 10S 9S 7S 9S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 10S 9S 8S 9S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 10S 9S 9S 9S 0S 0S 0S 0S ]
[0000]        STX[8] = [ 10S 9S 9S 9S 1S 0S 0S 0S ]
[0000]        STX[8] = [ 10S 9S 9S 9S 1S 1S 0S 0S ]
[0000]        STX[8] = [ 10S 9S 9S 9S 1S 1S 1S 0S ]
[0000]        STX[8] = [ 10S 9S 9S 9S 2S 1S 1S 0S ]
[0000]        STX[8] = [ 10S 9S 9S 9S 3S 1S 1S 0S ]
[0000]        STX[8] = [ 10S 9S 9S 9S 4S 1S 1S 0S ]
[0000]        STX[8] = [ 10S 9S 9S 9S 5S 1S 1S 0S ]
[0000]        STX[8] = [ 10S 9S 9S 9S 6S 1S 1S 0S ]
[0000]        STX[8] = [ 10S 9S 9S 9S 7S 1S 1S 0S ]
[0000]        STX[8] = [ 10S 9S 9S 9S 8S 1S 1S 0S ]
[0000]        STX[8] = [ 10S 9S 9S 9S 9S 1S 1S 0S ]
[0000]        STX[8] = [ 10S 9S 9S 9S 9S 1S 2S 0S ]
[0000]        STX[8] = [ 10S 9S 9S 9S 9S 1S 3S 0S ]
[0000]        STX[8] = [ 10S 9S 9S 9S 9S 1S 4S 0S ]
[0000]        STX[8] = [ 10S 9S 9S 9S 9S 1S 5S 0S ]
[0000]        STX[8] = [ 10S 9S 9S 9S 9S 1S 6S 0S ]
[0000]        STX[8] = [ 10S 9S 9S 9S 9S 1S 7S 0S ]
[0000]        STX[8] = [ 10S 9S 9S 9S 9S 1S 8S 0S ]
[0000]        STX[8] = [ 10S 9S 9S 9S 9S 1S 9S 0S ]
[0000]        STX[8] = [ 10S 9S 9S 9S 9S 2S 9S 0S ]
[0000]        STX[8] = [ 10S 9S 9S 9S 9S 3S 9S 0S ]
[0000]        STX[8] = [ 10S 9S 9S 9S 9S 4S 9S 0S ]
[0000]        STX[8] = [ 10S 9S 9S 9S 9S 5S 9S 0S ]
[0000]        STX[8] = [ 10S 9S 9S 9S 9S 6S 9S 0S ]
[0000]        STX[8] = [ 10S 9S 9S 9S 9S 7S 9S 0S ]
[0000]        STX[8] = [ 10S 9S 9S 9S 9S 8S 9S 0S ]
[0000]        STX[8] = [ 10S 9S 9S 9S 9S 9S 9S 0S ]
[0000]        STX[8] = [ 10S 9S 9S 9S 9S 9S 9S 1S ]
```

I can't quite figure out what's different.  Do you see it?

File: src/transport_ofi.c [hunk 1]
```diff
@@ -463,18 +420,27 @@ int shmem_transport_ofi_stx_search_shared(long threshold)
...
-                    (shmem_transport_ofi_stx_threshold_cmp(i, threshold) || threshold == -1) &&
-                    !shmem_transport_ofi_stx_pool[i].is_private) {
-                    stx_idx = i;
-                    rand_pool_prev_choice = stx_idx;
-                    rr_start_idx = (i + 1) % shmem_transport_ofi_stx_max;
-                    break;
+                    (shmem_transport_ofi_stx_pool[i].ref_cnt <= threshold || threshold == -1) &&
+                    !shmem_transport_ofi_stx_pool[i].is_private)
+                {
+                    ++count;
```
Comment (davidozog) [inline src/transport_ofi.c]: Do we want a `break;` here?

**Comment by davidozog** [comment]
Ok, maybe I'm just forgetting that the allocator scheme only applies after all STXs are filled to the threshold.  Up to the threshold, the scheme is always ~~round-robin~~ sequential.  Is that right?

In that case, it looks like my original code has a bug in it.

**Review by davidozog — APPROVED** [review]

File: src/transport_ofi.c (hunk 1 above)
Comment (jdinan) [reply src/transport_ofi.c]: Oh, yeah, that's a nice optimization to break early.  IIUC, this shouldn't change the behavior of the allocator (apart from making it faster, which is always a good thing), right?

**Comment by jdinan** [comment]
Yeah, if you only have one shared STX that's below the threshold, you will keep picking that one.  Until there are none below the threshold, at which point you allocate a new STX and repeat.

**Comment by jdinan** [comment]
BTW, you can merge these changes into your branch whenever you're ready.  Then they will show up on the SOS PR and we can put them through the usual testing again.

File: src/transport_ofi.c (hunk 1 above)
Comment (davidozog) [reply src/transport_ofi.c]: Right!

**Pull request merged**